server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -777,6 +813,8 @@ public void executeFetchPhase(ShardFetchRequest request, SearchShardTask task, A
         final ReaderContext readerContext = findReaderContext(request.contextId(), request);
         final ShardSearchRequest shardSearchRequest = readerContext.getShardSearchRequest(request.getShardSearchRequest());
         final Releasable markAsUsed = readerContext.markAsUsed(getKeepAlive(shardSearchRequest));
+        listener = tryFetchLookupFields(shardSearchRequest, task, listener);
```

Comment:
It would be really nice if we could handle this in the fetch phase (maybe even in `FieldFetcher`) instead of having special handling for it here. Did you choose not to do that because of concerns over blocking the search thread?

Reply:
That's correct. The FetchPhase is synchronous, while the lookup process is async. I tried to make the FetchPhase async, but it requires more work (mainly because aggregations also use it). I can back out this optimization (i.e., tries to look up using local nodes only) and introduce it later if you and Yannick prefer.